### ISSUE TITLE
How to automatically end training after reaching goal Mean Reward & Std of Reward?

### ISSUE THREAD

**Issue by bybgr** [Unity-Technologies/ml-agents#1030]
I wonder how to stop training immediately after the model reaches certain Mean Reward & Std of Reward? I currently use 'Ctrl+C' to manually stop training when I am happy about Mean Reward & Std of Reward. Is there any method to do that automatically?
Thank you!

**Comment by scotthovestadt** [comment]
@mbaske That's incredibly cool. I looked through the code but haven't set it up yet.

So, it runs for a full session, saves the training data to a JSON files, and then re-runs using the same training data with different hyper-parameters. Very, very cool. I'm definitely going to try it.

What is needed to have it as a pure add-on module to ml-agents instead of needing a separate version of `learn.py` and other files? Just a few hooks to save training data? Maybe can get a PR into the official repo to add any hooks needed to make it a pure add-on... more maintainable that way.

**Comment by mbaske** [comment]
@scotthovestadt My code runs learn.py in a subprocess and uses JSON files to exchange training data with it. Specifically hyperparameters and stop conditions when a trainer starts, training statistics when it fininshes.

I see your point regarding maintainability. But since this setup requires only a few changes to the ml-agents code, I decided against duplicating everything as an add-on. Besides, it's still possible to run learn.py directly with the usual options.

If you just need to save training stats, the write_summary method in https://github.com/Unity-Technologies/ml-agents/blob/master/python/unitytrainers/trainer.py should be a good place for implementing that.

**Comment by bybgr** [comment]
@mbaske Thank you so much! I will take a look.

**Comment by xiaomaogy** [comment]
Thanks for reaching out to us. Hopefully you were able to resolve your issue.  We are closing this due to inactivity, but if you need additional assistance, feel free to reopen the issue.
